File: DesktopSample/Advanced/PageAndRDL/OracleDataProvider/C#/TestDesignerPro/DesignerForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using System.IO;
﻿using ActiveReports.Samples.TestDesignerPro.Properties;
using GrapeCity.ActiveReports.Design;
using GrapeCity.ActiveReports;
using GrapeCity.ActiveReports.Viewer.Win.Internal.Export;
using ActiveReports.Samples.TestDesignerPro.Export;
using GrapeCity.Viewer.Common.ViewModel;
using GrapeCity.ActiveReports.PageReportModel;
using System.Linq;

namespace ActiveReports.Samples.TestDesignerPro
{
	public partial class DesignerForm : Form
	{
		private string _reportName = "DemoReport.rdlx";
		private ExportForm _exportForm;
		private ExportViewerFactory _exportViewerFactory;
		private ToolStripMenuItem _exportToolStrip;
		private ToolStripButton _exportToolStripButton;

		public DesignerForm()
		{		
			InitializeComponent();

			_exportForm = new ExportForm();
			_exportViewerFactory = new ExportViewerFactory();
			reportDesigner.DesignerTabChanged += ReportDesigner_DesignerTabChanged;
			reportDesigner.NewReport(DesignerReportType.Page);//デザイナで空のページレポートを設定します。

			//ツールボックス、レポートエクスプローラおよびプロパティグリッドに項目を追加します。
			reportDesigner.Toolbox = reportToolbox;//ツールボックスをレポートデザイナに接続します。
			reportDesigner.LayoutChanged += OnDesignerLayoutChanged;
			reportExplorer.ReportDesigner = reportDesigner;//レポートエクスプローラをレポートデザイナに接続します
			reportDesigner.PropertyGrid = propertyGrid;//プロパティグリッドをレポートデザイナにに接続します。
			groupEditor.ReportDesigner = reportDesigner;
			layerList.ReportDesigner = reportDesigner;
			reportsLibrary.ReportDesigner = reportDesigner;

			//メニューに項目を追加します。
			ToolStrip toolStrip = reportDesigner.CreateToolStrips(new DesignerToolStrips[]
				{
					DesignerToolStrips.Menu
				})[0];
			toolStrip.Items.RemoveAt(2);
			ToolStripDropDownItem fileMenu = (ToolStripDropDownItem)toolStrip.Items[0];
			CreateFileMenu(fileMenu);
			AppendToolStrips(0, new ToolStrip[]
				{
					toolStrip
				});
			AppendToolStrips(1, reportDesigner.CreateToolStrips(new DesignerToolStrips[]
				{
					DesignerToolStrips.Edit, 
					DesignerToolStrips.Undo, 
					DesignerToolStrips.Zoom
				}));
			AppendToolStrips(1, new ToolStrip[]
				{
					CreateReportToolbar()
				});
			AppendToolStrips(2, reportDesigner.CreateToolStrips(new DesignerToolStrips[]
				{
					DesignerToolStrips.Format, 
					DesignerToolStrips.Layout
				}));
			reportDesigner.ReportChanged += (_, __) => UpdateReportName();
			InitGroupEditorToggle();
			SetExportState(false);
		}

		private void ReportDesigner_DesignerTabChanged(object sender, EventArgs e)
		{
			if (reportDesigner == null || reportDesigner.ActiveTab == DesignerTab.Designer)
			{
				SetExportState(false);
				return;
			}

			bool enabledExport = reportDesigner.ActiveTab == DesignerTab.Preview && reportDesigner.ReportViewer != null;
			SetExportState(enabledExport);
		}

		private void SetExportState(bool state)
		{
			if (_exportToolStrip != null) _exportToolStrip.Enabled = state;
			if (_exportToolStripButton != null) _exportToolStripButton.Enabled = state;
		}

		private void Export()
		{
			if (_exportForm == null || _exportViewerFactory == null || reportDesigner.ActiveTab != DesignerTab.Preview)
				return;

			_exportForm.Show(this, _exportViewerFactory.CreateExportViewer(reportDesigner.ReportViewer), DetermineReportType());
		}

		private ExportForm.ReportType DetermineReportType()
		{
			if (reportDesigner.ActiveTab == DesignerTab.Preview)
			{
				return DetermineReportTypeByOpenedReport();
			}

			var sectionReport = reportDesigner.Report as SectionReport;
			if (sectionReport != null) return ExportForm.ReportType.Section;

			var pageReport = reportDesigner.Report as PageReport;
			if (pageReport == null) return DefaultReportType;

			var report = pageReport.Report;

			if (report?.ViewerType == ViewerType.Dashboard) return ExportForm.ReportType.Dashboard;

			if (report?.ReportSections?.FirstOrDefault()?.Body == null) return DefaultReportType;

			ReportItemCollection items = report.ReportSections.FirstOrDefault()?.Body.ReportItems;
			return items != null && items.Count == 1 && items[0] is FixedPage ? ExportForm.ReportType.PageFpl : ExportForm.ReportType.PageCpl;
		}

		private const ExportForm.ReportType DefaultReportType = ExportForm.ReportType.PageCpl;

		private ExportForm.ReportType DetermineReportTypeByOpenedReport()
		{
			OpenedReport openedReport = reportDesigner.ReportViewer.OpenedReport;

			switch (openedReport)
			{
				case OpenedReport.Fpl:
					return ExportForm.ReportType.PageFpl;
				case OpenedReport.Dashboard:
					return ExportForm.ReportType.Dashboard;
				case OpenedReport.Section:
					return ExportForm.ReportType.Section;
				case OpenedReport.Cpl:
				case OpenedReport.None:
				default:
					return DefaultReportType;
			}
		}

		private void OnExport(object sender, EventArgs e) => Export();

		private void SetReportName(string reportName)
		{
			if (string.IsNullOrEmpty(reportName))
			{
				reportDesigner.IsDirty = false;
				_reportName = reportDesigner.Report is PageReport ? Resources.DefaultReportNameRdlx : Resources.DefaultReportNameRpx;
			}
			else
			{
				_reportName = reportName;
			}

			Text = Resources.SampleNameTitle + Path.GetFileName(_reportName) + (reportDesigner.IsDirty ? Resources.DirtySign : string.Empty);
		}

		private void UpdateReportName()
		{
			SetReportName(_reportName);
		}

		private int _groupEditorSize;
		private void InitGroupEditorToggle()
		{
			GroupEditorToggleButton.Image = Resources.GroupEditorHide;
			GroupEditorToggleButton.MouseEnter += (sender, args) => { GroupEditorToggleButton.BackColor = Color.LightGray; };
			GroupEditorToggleButton.MouseLeave += (sender, args) => { GroupEditorToggleButton.BackColor = Color.Gainsboro; };
			GroupEditorToggleButton.Click += (sender, args) =>
			{
				if (groupEditor.Visible)
				{
					GroupEditorToggleButton.Image = Resources.GroupEditorShow;
					_groupEditorSize = splitContainer1.ClientSize.Height - splitContainer1.SplitterDistance;
					splitContainer1.SplitterDistance = splitContainer1.ClientSize.Height - GroupEditorSeparator.Height - GroupEditorContainer.Padding.Vertical - splitContainer1.Panel2.Padding.Vertical - splitContainer1.SplitterWidth;
					splitContainer1.IsSplitterFixed = true;
					groupEditor.Visible = false;
				}
				else
				{
					GroupEditorToggleButton.Image = Resources.GroupEditorHide;
					splitContainer1.SplitterDistance = _groupEditorSize < splitContainer1.ClientSize.Height ? splitContainer1.ClientSize.Height - _groupEditorSize : splitContainer1.ClientSize.Height * 2 / 3;
					splitContainer1.IsSplitterFixed = false;
					groupEditor.Visible = true;
				}
			};

			groupEditor.VisibleChanged += (sender, args) => GroupPanelVisibility.SetToolTip(GroupEditorToggleButton, groupEditor.Visible ? Resources.HideGroupPanelToolTip : Resources.ShowGroupPanelToolTip);
		}

		//ToolStripDropDownItemにDropDownItemsを追加します。
		private void CreateFileMenu(ToolStripDropDownItem fileMenu)
		{
			fileMenu.DropDownItems.Clear();
			fileMenu.DropDownItems.Add(new ToolStripMenuItem(Resources.MenuNew, Resources.CmdNewReport, new EventHandler(OnNew), (Keys.Control | Keys.N)));
			fileMenu.DropDownItems.Add(new ToolStripMenuItem(Resources.MenuOpen, Resources.CmdOpen, new EventHandler(OnOpen), (Keys.Control | Keys.O)));
			fileMenu.DropDownItems.Add(new ToolStripMenuItem(Resources.MenuSave, Resources.CmdSave, new EventHandler(OnSave), (Keys.Control | Keys.S)));
			fileMenu.DropDownItems.Add(new ToolStripMenuItem(Resources.MenuSaveAs, Resources.CmdSaveAs, new EventHandler(OnSaveAs)));
			fileMenu.DropDownItems.Add(new ToolStripSeparator());
			_exportToolStrip = new ToolStripMenuItem(Resources.Export, Resources.CmdExport, new EventHandler(OnExport), (Keys.Control | Keys.E));
			fileMenu.DropDownItems.Add(_exportToolStrip);
			fileMenu.DropDownItems.Add(new ToolStripSeparator());
			fileMenu.DropDownItems.Add(new ToolStripMenuItem(Resources.MenuExit, null, new EventHandler(OnExit)));
		}

		private ToolStrip CreateReportToolbar()
		{
			_exportToolStripButton = CreateToolStripButton(Resources.Export, Resources.CmdExport, new EventHandler(OnExport), Resources.Export);

			return new ToolStrip(new ToolStripButton[]
			{
				CreateToolStripButton(Resources.MenuNew, Resources.CmdNewReport,new EventHandler(OnNew), Resources.MenuNew),
				CreateToolStripButton(Resources.MenuOpen, Resources.CmdOpen,new EventHandler(OnOpen), Resources.MenuOpen),
				CreateToolStripButton(Resources.MenuSave, Resources.CmdSave,new EventHandler(OnSave), Resources.MenuSave),
				_exportToolStripButton
			});
		}

		//デザイナで「新規作成」をクリックして、新しいレポートを開きます。
		private void OnNew(object sender, EventArgs e)
		{
			if (ConfirmSaveChanges())
			{
				reportDesigner.ReportChanged -= (_, __) => UpdateReportName();
				reportDesigner.ExecuteAction(DesignerAction.NewReport);
				reportDesigner.IsDirty = false;
				SetReportName(null);
				reportDesigner.ReportChanged += (_, __) => UpdateReportName();
			}
			ShowHideGroupEditor();
		}

		//デザイナで「開く」をクリックして、新しいレポートを開きます。
		private void OnOpen(object sender, EventArgs e)
		{
			if (!ConfirmSaveChanges())
			{
				return;
			}

			using (var openDialog = new OpenFileDialog())
			{
				openDialog.FileName = string.Empty;
				openDialog.Filter = Properties.Resources.RdlxFilter;
				if (openDialog.ShowDialog(this) == DialogResult.OK)
				{
					_reportName = openDialog.FileName;
					reportDesigner.LoadReport(new FileInfo(_reportName));
				}
			}
			ShowHideGroupEditor();
		}

		private void OnDesignerLayoutChanged(object sender, LayoutChangedArgs e)
		{
			// load or new report events
			if (e.Type == LayoutChangeType.ReportLoad || e.Type == LayoutChangeType.ReportClear)
			{
				reportToolbox.Reorder(reportDesigner);
				reportToolbox.EnsureCategories(); //check Data tools availability
				reportToolbox.Refresh();
			}
		}

		private void ShowHideGroupEditor()
		{
			if (reportDesigner.ReportType == DesignerReportType.Section)
			{
				splitContainer1.Panel2Collapsed = true;
			}
			else
			{
				splitContainer1.Panel2Collapsed = false;
			}
		}

		//デザイナで「上書きを保存」をクリックして、新しいレポートを開きます。
		private void OnSave(object sender, EventArgs e)
		{
			if (string.IsNullOrEmpty(_reportName)
				|| string.IsNullOrEmpty(Path.GetDirectoryName(_reportName))
				|| !File.Exists(_reportName))
			{
				if (PerformSaveAs())
					reportDesigner.SaveReport(new FileInfo(_reportName));
			}
			else
			{
				reportDesigner.SaveReport(new FileInfo(_reportName));
			}
			SetReportName(_reportName);
		}

		//デザイナで「名前を付けて保存」をクリックして、新しいレポートを開きます。
		private void OnSaveAs(object sender, EventArgs e)
		{
			PerformSaveAs();
		}

		private bool PerformSaveAs()
		{
			using (var saveDialog = new SaveFileDialog())
			{
				saveDialog.Filter = GetSaveFilter();
				saveDialog.FileName = Path.GetFileName(_reportName);
				saveDialog.DefaultExt = ".rdlx";
				saveDialog.InitialDirectory = new DirectoryInfo(Application.ExecutablePath).Parent.Parent.Parent.FullName;
				if (saveDialog.ShowDialog() == DialogResult.OK)
				{
					_reportName = saveDialog.FileName;
					reportDesigner.SaveReport(new FileInfo(_reportName));
					reportDesigner.IsDirty = false;
					return true;
				}
			}

			return false;
		}

		private string GetSaveFilter()
		{
			switch (reportDesigner.ReportType)
			{
				case DesignerReportType.Section:
					return Resources.RpxFilter;
				case DesignerReportType.Page:
				case DesignerReportType.Rdl:
					return Resources.RdlxFilter;
				default:
					return Resources.RpxFilter;
			}
		}

		private void OnExit(object sender, EventArgs e)
		{
			Close();
		}

		//修正がデザイナーに読み込まれたレポートに加えられたかどうかをチェックする
		private bool ConfirmSaveChanges()
		{
			if (reportDesigner.IsDirty)
			{
				DialogResult result = MessageBox.Show(this, Resources.SaveConformation, "", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question);

				if (result == DialogResult.Cancel)
				{
					return false;
				}
				if (result == DialogResult.Yes)
				{
					return PerformSaveAs();
				}
			}
			return true;
		}
	   
		private void AppendToolStrips(int row, IList<ToolStrip> toolStrips)
		{
			ToolStripPanel topToolStripPanel = toolStripContainer.TopToolStripPanel;
			int num = toolStrips.Count;
			while (--num >= 0)
			{
				topToolStripPanel.Join(toolStrips[num], row);
			}
		}

		private static ToolStripButton CreateToolStripButton(string text, System.Drawing.Image image, EventHandler handler, string toolTip)
		{
			return new ToolStripButton(text, image, handler)
			{
				DisplayStyle = ToolStripItemDisplayStyle.Image,
				ToolTipText = toolTip,
				DoubleClickEnabled = true
			};
		}

		private void UnifiedDesignerForm_Load(object sender, EventArgs e)
		{
			reportDesigner.LoadReport(new FileInfo(_reportName));
		}

		private void CreateReportExplorer()
		{
			if (reportDesigner.ReportType == DesignerReportType.Section)
			{
				if (explorerpropertygridContainer.Panel1.Controls.Contains(reportExplorerTabControl))
				{
					reportExplorerTabControl.TabPages[0].SuspendLayout();
					explorerpropertygridContainer.Panel1.SuspendLayout();
					explorerpropertygridContainer.Panel1.Controls.Remove(reportExplorerTabControl);
					explorerpropertygridContainer.Panel1.Controls.Add(reportExplorer);
					reportExplorerTabControl.TabPages[0].ResumeLayout();
					explorerpropertygridContainer.Panel1.ResumeLayout();
				}
			}
			else if (!explorerpropertygridContainer.Panel1.Controls.Contains(reportExplorerTabControl))
			{
				reportExplorerTabControl.TabPages[0].SuspendLayout();
				explorerpropertygridContainer.Panel1.SuspendLayout();
				explorerpropertygridContainer.Panel1.Controls.Remove(reportExplorer);
				reportExplorerTabControl.TabPages[0].Controls.Add(reportExplorer);
				explorerpropertygridContainer.Panel1.Controls.Add(reportExplorerTabControl);
				reportExplorerTabControl.TabPages[0].ResumeLayout();
				explorerpropertygridContainer.Panel1.ResumeLayout();
			}
		}

		private void reportDesigner_LayoutChanged(object sender, LayoutChangedArgs e)
		{
			CreateReportExplorer();
		}
	}
}

File: DesktopSample/Advanced/PageAndRDL/OracleDataProvider/VB.NET/TestDesignerPro/DesignerForm.vb
﻿Imports GrapeCity.ActiveReports
Imports GrapeCity.ActiveReports.Design
Imports GrapeCity.ActiveReports.PageReportModel
Imports GrapeCity.ActiveReports.Viewer.Win.Internal.Export
Imports GrapeCity.Viewer.Common.ViewModel
Imports System.IO
Imports System.Text

Public Class DesignerForm
	Private _reportName As String = "DemoReport.rdlx"
	Private _exportForm As ExportForm
	Private _exportViewerFactory As ExportViewerFactory
	Private _exportToolStrip As ToolStripMenuItem
	Private _exportToolStripButton As ToolStripButton

	Private Sub UnifiedDesignerForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
		reportDesigner.LoadReport(New FileInfo(_reportName))
	End Sub

	Public Sub New()
		Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
		InitializeComponent()

		_exportForm = New ExportForm
		_exportViewerFactory = New ExportViewerFactory
		AddHandler reportDesigner.DesignerTabChanged, AddressOf ReportDesigner_DesignerTabChanged

		reportDesigner.NewReport(DesignerReportType.Page) 'Sets the designer to create a blank page based report
		'Populating the ToolBox, ReportExplorer and PropertyGrid.
		reportDesigner.Toolbox = reportToolbox 'Attaches the toolbox to the report designer
		AddHandler reportDesigner.LayoutChanged, Sub(s, e) OnDesignerLayoutChanged(s, e)
		reportExplorer.ReportDesigner = reportDesigner 'Attaches the report explorer to the report designer
		groupEditor.ReportDesigner = reportDesigner
		reportsLibrary.ReportDesigner = reportDesigner
		layerList.ReportDesigner = reportDesigner
		reportDesigner.PropertyGrid = propertyGrid 'Attaches the Property Grid to the report designer
		'Populating the menu.
		Dim toolStrip As ToolStrip = reportDesigner.CreateToolStrips(New DesignerToolStrips() {DesignerToolStrips.Menu})(0)
		toolStrip.Items.RemoveAt(2)
		Dim fileMenu As ToolStripDropDownItem = DirectCast(toolStrip.Items(0), ToolStripDropDownItem)
		CreateFileMenu(fileMenu)
		AppendToolStrips(0, New ToolStrip() {toolStrip})
		AppendToolStrips(1, reportDesigner.CreateToolStrips(New DesignerToolStrips() {DesignerToolStrips.Edit, DesignerToolStrips.Undo, DesignerToolStrips.Zoom}))
		AppendToolStrips(1, New ToolStrip() {CreateReportToolbar()})
		AppendToolStrips(2, reportDesigner.CreateToolStrips(New DesignerToolStrips() {DesignerToolStrips.Format, DesignerToolStrips.Layout}))
		AddHandler reportDesigner.ReportChanged, AddressOf UpdateReportName
		InitGroupEditorToggle()
		SetExportState(False)
	End Sub

	Private Sub ReportDesigner_DesignerTabChanged(o As Object, e As EventArgs)
		If reportDesigner Is Nothing Or reportDesigner.ActiveTab = DesignerTab.Designer Then
			SetExportState(False)
			Return
		End If

		Dim enabledExport As Boolean = reportDesigner.ActiveTab = DesignerTab.Preview And Not (reportDesigner.ReportViewer Is Nothing)
		SetExportState(enabledExport)
	End Sub

	Private Sub SetExportState(state As Boolean)
		If Not (_exportToolStrip Is Nothing) Then _exportToolStrip.Enabled = state
		If Not (_exportToolStripButton Is Nothing) Then _exportToolStripButton.Enabled = state
	End Sub

	Private Sub Export()
		If _exportForm Is Nothing Or _exportViewerFactory Is Nothing Or Not (reportDesigner.ActiveTab = DesignerTab.Preview) Then
			Return
		End If

		_exportForm.Show(Me, _exportViewerFactory.CreateExportViewer(reportDesigner.ReportViewer), DetermineReportType())
	End Sub

	Private Function DetermineReportType() As ExportForm.ReportType
		If Not (reportDesigner.ActiveTab = DesignerTab.Preview) Then
			Return DetermineReportTypeByOpenedReport()
		End If

		Dim sectionReport As SectionReport = TryCast(reportDesigner.Report, SectionReport)
		If Not (sectionReport Is Nothing) Then
			Return ExportForm.ReportType.Section
		End If

		Dim pageReport As PageReport = TryCast(reportDesigner.Report, PageReport)
		If Not (pageReport Is Nothing) Then
			Return DefaultReportType
		End If

		Dim report As Report = pageReport.Report

		If report?.ViewerType = ViewerType.Dashboard Then
			Return ExportForm.ReportType.Dashboard
		End If

		If report?.ReportSections?.FirstOrDefault()?.Body Is Nothing Then
			Return DefaultReportType
		End If

		Dim items As ReportItemCollection = report.ReportSections.FirstOrDefault()?.Body.ReportItems

		If items?.Count = 1 And items(0).GetType() Is GetType(FixedPage) Then
			Return ExportForm.ReportType.PageFpl
		Else
			Return ExportForm.ReportType.PageCpl
		End If
	End Function

	Private Const DefaultReportType As ExportForm.ReportType = ExportForm.ReportType.PageCpl

	Private Function DetermineReportTypeByOpenedReport() As ExportForm.ReportType
		Dim openedReport As OpenedReport = reportDesigner.ReportViewer.OpenedReport

		Select Case openedReport
			Case OpenedReport.Fpl
				Return ExportForm.ReportType.PageFpl
			Case OpenedReport.Dashboard
				Return ExportForm.ReportType.Dashboard
			Case OpenedReport.Section
				Return ExportForm.ReportType.Section
			Case OpenedReport.Cpl
			Case OpenedReport.None
				Return DefaultReportType
		End Select

		Return DefaultReportType
	End Function

	Private Sub OnExport(o As Object, e As EventArgs)
		Export()
	End Sub

	Private Sub SetReportName(reportName As String)
		If String.IsNullOrEmpty(reportName) Then
			_reportName = If(TypeOf reportDesigner.Report Is PageReport, My.Resources.DefaultReportNameRdlx, My.Resources.DefaultReportNameRpx)
		Else
			_reportName = reportName
		End If
		Text = My.Resources.SampleNameTitle + Path.GetFileName(_reportName) + (If(reportDesigner.IsDirty, My.Resources.DirtySign, String.Empty))
	End Sub

	Private Sub UpdateReportName(o As Object, e As EventArgs)
		SetReportName(_reportName)
	End Sub

	Private _groupEditorSize As Integer
	Private Sub InitGroupEditorToggle()
		GroupEditorToggleButton.Image = My.Resources.GroupEditorHide
		AddHandler GroupEditorToggleButton.MouseEnter, Sub(sender, args)
														   GroupEditorToggleButton.BackColor = Color.LightGray

													   End Sub
		AddHandler GroupEditorToggleButton.MouseLeave, Sub(sender, args)
														   GroupEditorToggleButton.BackColor = Color.Gainsboro

													   End Sub
		AddHandler GroupEditorToggleButton.Click, Sub(sender, args)
													  If groupEditor.Visible Then
														  GroupEditorToggleButton.Image = My.Resources.GroupEditorShow
														  _groupEditorSize = SplitContainer1.ClientSize.Height - SplitContainer1.SplitterDistance
														  SplitContainer1.SplitterDistance = SplitContainer1.ClientSize.Height - GroupEditorSeparator.Height - GroupEditorContainer.Padding.Vertical - SplitContainer1.Panel2.Padding.Vertical - SplitContainer1.SplitterWidth
														  SplitContainer1.IsSplitterFixed = True
														  groupEditor.Visible = False
													  Else
														  GroupEditorToggleButton.Image = My.Resources.GroupEditorHide
														  SplitContainer1.SplitterDistance = CInt(If(_groupEditorSize < SplitContainer1.ClientSize.Height, SplitContainer1.ClientSize.Height - _groupEditorSize, SplitContainer1.ClientSize.Height * 2 / 3))
														  SplitContainer1.IsSplitterFixed = False
														  groupEditor.Visible = True
													  End If

												  End Sub

		AddHandler groupEditor.VisibleChanged, Sub() GroupPanelVisibility.SetToolTip(GroupEditorToggleButton, If(groupEditor.Visible, My.Resources.HideGroupPanelToolTip, My.Resources.ShowGroupPanelToolTip))
	End Sub

	'Adding DropDownItems to the ToolStripDropDownItem
	Private Sub CreateFileMenu(ByVal fileMenu As ToolStripDropDownItem)
		fileMenu.DropDownItems.Clear()
		fileMenu.DropDownItems.Add(New ToolStripMenuItem(My.Resources.MenuNew, My.Resources.CmdNewReport, New EventHandler(AddressOf OnNew), (Keys.Control Or Keys.N)))
		fileMenu.DropDownItems.Add(New ToolStripMenuItem(My.Resources.MenuOpen, My.Resources.CmdOpen, New EventHandler(AddressOf OnOpen), (Keys.Control Or Keys.O)))
		fileMenu.DropDownItems.Add(New ToolStripMenuItem(My.Resources.MenuSave, My.Resources.CmdSave, New EventHandler(AddressOf OnSave), (Keys.Control Or Keys.S)))
		fileMenu.DropDownItems.Add(New ToolStripMenuItem(My.Resources.MenuSaveAs, My.Resources.CmdSaveAs, New EventHandler(AddressOf OnSaveAs)))
		fileMenu.DropDownItems.Add(New ToolStripSeparator())
		_exportToolStrip = New ToolStripMenuItem(My.Resources.Export, My.Resources.CmdExport, New EventHandler(AddressOf OnExport), (Keys.Control Or Keys.E))
		fileMenu.DropDownItems.Add(_exportToolStrip)
		fileMenu.DropDownItems.Add(New ToolStripSeparator())
		fileMenu.DropDownItems.Add(New ToolStripMenuItem(My.Resources.MenuExit, Nothing, New EventHandler(AddressOf OnExit)))
	End Sub

	'Getting the Designer to open a new report on "New" menu item click.
	Private Sub OnNew(ByVal sender As Object, ByVal e As EventArgs)
		If ConfirmSaveChanges() Then
			reportDesigner.IsDirty = False
			RemoveHandler reportDesigner.ReportChanged, AddressOf UpdateReportName
			reportDesigner.ExecuteAction(DesignerAction.NewReport)
			SetReportName(Nothing)
			AddHandler reportDesigner.ReportChanged, AddressOf UpdateReportName
			ShowHideGroupEditor()
		End If
	End Sub

	'Getting the Designer to open a report on "Open" menu item click.
	Private Sub OnOpen(ByVal sender As Object, ByVal e As EventArgs)
		If Not ConfirmSaveChanges() Then
			Return
		End If

		Using openDialog = New OpenFileDialog()
			openDialog.FileName = String.Empty
			openDialog.Filter = My.Resources.RdlxFilter
			If openDialog.ShowDialog(Me) = DialogResult.OK Then
				_reportName = openDialog.FileName
				reportDesigner.LoadReport(New FileInfo(_reportName))
			End If
		End Using

		ShowHideGroupEditor()
	End Sub

	Private Sub OnDesignerLayoutChanged(sender As Object, e As LayoutChangedArgs)
		If e.Type = LayoutChangeType.ReportLoad OrElse e.Type = LayoutChangeType.ReportClear Then
			reportToolbox.Reorder(reportDesigner)
			reportToolbox.EnsureCategories()
			reportToolbox.Refresh()
		End If
	End Sub

	'Getting the Designer to open a report on "Save" menu item click.
	Private Sub OnSave(ByVal sender As Object, ByVal e As EventArgs)
		If String.IsNullOrEmpty(_reportName) OrElse String.IsNullOrEmpty(Path.GetDirectoryName(_reportName)) OrElse Not File.Exists(_reportName) Then
			If PerformSaveAs() Then
				reportDesigner.SaveReport(New FileInfo(_reportName))
			End If
		Else
			reportDesigner.SaveReport(New FileInfo(_reportName))
		End If
		SetReportName(_reportName)
	End Sub

	Private Sub ShowHideGroupEditor()
		If reportDesigner.ReportType = DesignerReportType.Section Then
			SplitContainer1.Panel2Collapsed = True
		Else
			SplitContainer1.Panel2Collapsed = False
		End If
	End Sub

	'Getting the Designer to open a report on "Save As" menu item click.
	Private Sub OnSaveAs(ByVal sender As Object, ByVal e As EventArgs)
		PerformSaveAs()
	End Sub

	Private Function PerformSaveAs() As Boolean
		Using saveDialog = New SaveFileDialog()
			saveDialog.Filter = GetSaveFilter()
			saveDialog.FileName = Path.GetFileName(_reportName)
			saveDialog.DefaultExt = ".rdlx"
			saveDialog.InitialDirectory = New DirectoryInfo(Application.ExecutablePath).Parent.Parent.Parent.FullName
			If saveDialog.ShowDialog() = DialogResult.OK Then
				_reportName = saveDialog.FileName
				reportDesigner.SaveReport(New FileInfo(_reportName))
				reportDesigner.IsDirty = False
				Return True
			End If
		End Using

		Return False
	End Function

	Private Function GetSaveFilter() As String
		Select Case reportDesigner.ReportType
			Case DesignerReportType.Section
				Return My.Resources.RpxFilter
			Case DesignerReportType.Page, DesignerReportType.Rdl
				Return My.Resources.RdlxFilter
			Case Else
				Return My.Resources.RpxFilter
		End Select
	End Function

	Private Sub OnExit(ByVal sender As Object, ByVal e As EventArgs)
		Close()
	End Sub

	'Checking whether modifications have been made to the report loaded to the designer
	Private Function ConfirmSaveChanges() As Boolean
		If reportDesigner.IsDirty Then
			Dim result As DialogResult = MessageBox.Show(My.Resources.SaveConformation, "", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question)

			If result = DialogResult.Cancel Then
				Return False
			End If
			If result = DialogResult.Yes Then

				Return PerformSaveAs()
			End If
		End If
		Return True
	End Function

	Private Sub AppendToolStrips(ByVal row As Integer, ByVal toolStrips As IList(Of ToolStrip))
		Dim topToolStripPanel As ToolStripPanel = toolStripContainer.TopToolStripPanel
		Dim num As Integer = toolStrips.Count
		While Threading.Interlocked.Decrement(num) >= 0
			topToolStripPanel.Join(toolStrips(num), row)
		End While
	End Sub

	Private Function CreateReportToolbar() As ToolStrip
		_exportToolStripButton = CreateToolStripButton(My.Resources.Export, My.Resources.CmdExport, New EventHandler(AddressOf OnExport), My.Resources.Export)

		Return New ToolStrip(New ToolStripButton() {
							 CreateToolStripButton(My.Resources.MenuNew, My.Resources.CmdNewReport, New EventHandler(AddressOf OnNew), My.Resources.MenuNew),
							 CreateToolStripButton(My.Resources.MenuOpen, My.Resources.CmdOpen, New EventHandler(AddressOf OnOpen), My.Resources.MenuOpen),
							 CreateToolStripButton(My.Resources.MenuSave, My.Resources.CmdSave, New EventHandler(AddressOf OnSave), My.Resources.MenuSave),
							 _exportToolStripButton
							 })
	End Function

	Private Shared Function CreateToolStripButton(ByVal text As String, ByVal image As System.Drawing.Image, ByVal handler As EventHandler, ByVal toolTip As String) As ToolStripButton
		Return New ToolStripButton(text, image, handler) With {
		 .DisplayStyle = ToolStripItemDisplayStyle.Image,
			.ToolTipText = toolTip,
		 .DoubleClickEnabled = True
		}
	End Function

	Private Sub CreateReportExplorer()
		If reportDesigner.ReportType = DesignerReportType.Section Then

			If explorerpropertygridContainer.Panel1.Controls.Contains(reportExplorerTabControl) Then
				reportExplorerTabControl.TabPages(0).SuspendLayout()
				explorerpropertygridContainer.Panel1.SuspendLayout()
				explorerpropertygridContainer.Panel1.Controls.Remove(reportExplorerTabControl)
				explorerpropertygridContainer.Panel1.Controls.Add(reportExplorer)
				reportExplorerTabControl.TabPages(0).ResumeLayout()
				explorerpropertygridContainer.Panel1.ResumeLayout()
			End If
		ElseIf (Not explorerpropertygridContainer.Panel1.Controls.Contains(reportExplorerTabControl)) Then
			reportExplorerTabControl.TabPages(0).SuspendLayout()
			explorerpropertygridContainer.Panel1.SuspendLayout()
			explorerpropertygridContainer.Panel1.Controls.Remove(reportExplorer)
			reportExplorerTabControl.TabPages(0).Controls.Add(reportExplorer)
			explorerpropertygridContainer.Panel1.Controls.Add(reportExplorerTabControl)
			reportExplorerTabControl.TabPages(0).ResumeLayout()
			explorerpropertygridContainer.Panel1.ResumeLayout()
		End If
	End Sub

End Class
